File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/data/api/ErrorConverter.java
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.data.api;

import java.io.IOException;
import java.lang.annotation.Annotation;

import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

/**
 * Created by MateuszW on 2018-04-02.
 */
public class ErrorConverter {

    private final Retrofit retrofit;

    public ErrorConverter(Retrofit retrofit) {
        this.retrofit = retrofit;
    }

    public ErrorResponse convert(ResponseBody responseBody) {
        Converter<ResponseBody,ErrorResponse> converter = retrofit.responseBodyConverter(ErrorResponse.class, new Annotation[]{});
        try {
            return converter.convert(responseBody);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/data/api/ErrorConverter.kt
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.data.api

import okhttp3.ResponseBody
import retrofit2.Retrofit
import java.io.IOException

class ErrorConverter(private val retrofit: Retrofit) {

    fun convert(responseBody: ResponseBody): ErrorResponse? {
        val converter = retrofit.responseBodyConverter<ErrorResponse>(ErrorResponse::class.java, arrayOf())
        return try {
            converter.convert(responseBody)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }

    }
}
